File: jput-core/src/test/java/cz/encircled/jput/test/StatisticsTest.java
package cz.encircled.jput.test;

import cz.encircled.jput.Statistics;
import org.junit.Assert;
import org.junit.Test;

/**
 * @author Vlad on 27-May-17.
 */
public class StatisticsTest {

    @Test
    public void testAverage() {
        Assert.assertEquals(2, Statistics.getAverage(1, 3), 0);
        Assert.assertEquals(6, Statistics.getAverage(4, 6, 8), 0);
        Assert.assertEquals(98.5, Statistics.getAverage(100, 96, 99, 99), 0);
    }

    @Test
    public void testVariance() {
        Assert.assertEquals(1.0, Statistics.getVariance(1, 3), 0);
    }

    @Test
    public void testDeviation() {
        Assert.assertEquals(2.0, Statistics.getStandardDeviation(1, 5), 0);
    }

    @Test(expected = IllegalStateException.class)
    public void testArrayPercentileValid() {
        Statistics.getPercentile(new long[]{1, 10}, 1.01);
    }

    @Test(expected = IllegalStateException.class)
    public void testArrayPercentileValidLower() {
        Statistics.getPercentile(new long[]{1, 10}, -0.1);
    }

    @Test
    public void testArrayPercentile() {
        Assert.assertArrayEquals(new long[]{1, 2, 3, 4, 5, 6, 7, 8, 9}, Statistics.getPercentile(new long[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 0.9));
    }

    @Test
    public void testArrayPercentileRound() {
        Assert.assertArrayEquals(new long[]{1, 2, 3, 4, 5, 6, 7, 8, 9}, Statistics.getPercentile(new long[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 0.94));
        Assert.assertArrayEquals(new long[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, Statistics.getPercentile(new long[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}, 0.95));
    }

}

File: jput-core/src/test/java/cz/encircled/jput/test/StatisticsTest.kt
package cz.encircled.jput.test

import cz.encircled.jput.deviation
import cz.encircled.jput.percentile
import cz.encircled.jput.variance
import org.junit.Assert
import org.junit.Test
import kotlin.test.assertEquals

/**
 * @author Vlad on 27-May-17.
 */
class StatisticsTest {

    @Test
    fun testVariance() {
        Assert.assertEquals(4.0, listOf(1L, 5).variance(), 0.0)
        Assert.assertEquals(1.25, listOf(2L, 3, 4, 5).variance(), 0.0)
        Assert.assertEquals(200.0, listOf(50L, 60, 70, 80, 90).variance(), 0.0)
    }

    @Test
    fun testDeviation() {
        Assert.assertEquals(2.0, listOf(1L, 5).deviation(), 0.0)
    }

    @Test(expected = IllegalStateException::class)
    fun testArrayPercentileValid() {
        listOf(1L, 10).percentile(1.01)
    }

    @Test(expected = IllegalStateException::class)
    fun testArrayPercentileValidLower() {
        listOf(1L, 10).percentile(-0.1)
    }

    @Test
    fun testArrayPercentile() {
        val res = listOf(1L, 2, 3, 4, 5, 6, 7, 8, 9, 10).percentile(0.9)
        assertEquals(listOf<Long>(1, 2, 3, 4, 5, 6, 7, 8, 9), res)
    }

    @Test
    fun testArrayPercentileRound() {
        assertEquals(listOf<Long>(1, 2, 3, 4, 5, 6, 7, 8, 9), listOf(1L, 2, 3, 4, 5, 6, 7, 8, 9, 10).percentile(0.94))
        assertEquals(listOf<Long>(1, 2, 3, 4, 5, 6, 7, 8, 9, 10), listOf(1L, 2, 3, 4, 5, 6, 7, 8, 9, 10).percentile(0.95))
    }

}
